### ISSUE TITLE
DirectGpu OpenCL gemm_small + CLBlast-direct kernels produce numerically wrong results on AMD gfx1012 (RDNA1 Navi 14)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#364]
## Summary

`AiDotNet.Tensors.Engines.DirectGpuTensorEngine.TensorMatMul<float>` produces **numerically incorrect** results on AMD `gfx1012:xnack-` (Radeon RX 5500 / RDNA1 Navi 14) on Windows. The bug reproduces on both kernel paths the dispatcher selects: `gemm_small` for `M<128` (or `N<128` or `K<64`), and CLBlast-direct / `gemm_double_buffered` for `M,N>=128, K>=64`.

This silently corrupts `model.Train` / `model.Predict` for any HarmonicEngine consumer who turns on `AiModelBuilder.ConfigureGpuAcceleration()` — the marketing copy promises 10-100x speedups with no code change, but on this hardware the GPU returns garbage. Filed at upstream P0 against the `ConfigureGpuAcceleration` user contract.

## Environment

| field | value |
|---|---|
| AiDotNet | 0.198.0 |
| AiDotNet.Tensors | 0.81.0 |
| AiDotNet.Native.OpenBLAS | 0.81.0 |
| OS | Windows 10.0.26220 |
| Framework | .NET 10.0.8 |
| GPU | AMD Radeon RX 5500 (`gfx1012:xnack-`, RDNA1 Navi 14) |
| OpenCL runtime | Available (Native GPU libs: OpenCL=True) |
| CLBlast native | NOT loaded (`HasClBlast=False`) — so the suspect path is the in-tree `CLBlast direct kernels` (compiled from `.cl` source at startup, NOT the native CLBlast library) |
| CPU | x64 32-thread, AVX2+FMA, no AVX-512 |
| Autotune fingerprint | `x64-amd-avx2-cpu32` |
| Default deterministic mode | True (set by `AiModelBuilder.BuildAsync`; doesn't gate the DirectGpu path) |

`AiDotNet.Diagnostics.AccelerationDiagnostics.GetReport()` at startup:

```
=== AiDotNet Acceleration Report ===
  Engine:           Direct GPU Engine (OpenCL gfx1012:xnack-)
  Deterministic:    True
  ...
  GPU backends:     CUDA=False OpenCL=False HIP=False
  Native BLAS:      OpenBLAS=False CLBlast=False MKL=False CpuBLAS=False
  Native GPU libs:  CUDA=False HIP=False OpenCL=True
```

(Diagnostic divergence: `PlatformCapabilities.HasOpenCLSupport=False` while `NativeLibraryDetector.Status.HasOpenCl=True` — both come from `AccelerationDiagnostics`. Probably a separate filed issue, but worth flagging: the two probes disagree about whether OpenCL is available even though the engine is sitting on a fully-loaded `DirectGpuTensorEngine`.)

## Reproducer

```csharp
using AiDotNet.Tensors.Engines;
using AiDotNet.Tensors.LinearAlgebra;

// Build deterministic [64,64] x [64,64] float tensors.
var rng = new Random(42);
var a = new Tensor<float>([64, 64]);
var b = new Tensor<float>([64, 64]);
for (int i = 0; i < 64; i++)
    for (int j = 0; j < 64; j++) {
        a[i, j] = (float)(rng.NextDouble() - 0.5);
        b[i, j] = (float)(rng.NextDouble() - 0.5);
    }

// CPU baseline.
AiDotNetEngine.ResetToCpu();
var cpuResult = AiDotNetEngine.Current.TensorMatMul(a, b);
// cpuResult[0,0] ~ 0.458285

// DirectGpu.
using var gpuEngine = new DirectGpuTensorEngine();
Console.WriteLine($"isAvailable={gpuEngine.IsGpuAvailable}, name={gpuEngine.Name}");
// isAvailable=True, name=Direct GPU Engine (OpenCL gfx1012:xnack-)

AiDotNetEngine.Current = gpuEngine;
var gpuResult = AiDotNetEngine.Current.TensorMatMul(a, b);
_ = gpuResult.AsSpan();  // force deferred download to fire

Console.WriteLine($"cpu[0,0]={cpuResult[0,0]:G6}  gpu[0,0]={gpuResult[0,0]:G6}");
// cpu[0,0]=0.458285  gpu[0,0]=0.160251  (varies seed-to-seed but consistent within a run)

double maxErr = 0;
for (int i = 0; i < 64; i++)
    for (int j = 0; j < 64; j++) {
        double e = Math.Abs((double)cpuResult[i,j] - (double)gpuResult[i,j]);
        if (e > maxErr) maxErr = e;
    }
Console.WriteLine($"max_abs_err = {maxErr:G6}");  // ~ 3.18
```

### Phase 2b — `[64,64] x [64,64]` (`gemm_small` kernel path)

```
Pre-materialize gpuResult[0,0] = 0.160251
Post-AsSpan() gpuResult[0,0]   = 0.160251
GPU TensorMatMul: 89.11ms, max_abs_err vs CPU = 3.17851

Top 5 abs-error positions (i, j, gpu, cpu, |err|):
  [31,32]: gpu=0.0984512 cpu=-3.08006 err=3.17851
  [31,15]: gpu=0.402722  cpu=-2.12435 err=2.52707
  [30,53]: gpu=-0.126621 cpu=2.35923  err=2.48586
  [15,12]: gpu=-0.0684334 cpu=2.28911 err=2.35754
  [44,10]: gpu=-0.13121  cpu=2.21485  err=2.34606

GPU result first row (8 entries): 0.160251, 0.0951236, -0.208234, 0.221463, -0.103071, -0.0288194, 0.142358, -0.0555886
CPU result first row (8 entries): 0.458285, -0.4876, -0.107915, 0.139458, -0.683643, -0.10098, -0.843341, 0.527958
```

GPU values are roughly 3x smaller in magnitude than CPU — not random noise, not uninitialized memory, but a **systematic mis-accumulation**.

### Phase 2c — `[128,128] x [128,128]` (CLBlast-direct path)

```
Phase 2c: 128x128 GPU matmul max_abs_err vs CPU = inf
GPU [0,0] = -2.37012E+37  CPU [0,0] = -1.87171
```

Different kernel path, same broken outcome. The values overflow to ~1e+37 here because the broken accumulation now runs over 128 K elements instead of 64.

## What is NOT the cause

- **Not uninitialized memory** — values are deterministic across runs (same `0.160251` at `[0,0]` on rerun with same seed).
- **Not a missing deferred-download** — the `DeferredArrayMaterializer` registers correctly and fires on first indexer access (verified by reading `gpuResult[0,0]` both before and after explicit `AsSpan()` — identical result).
- **Not a `DownloadBuffer` sync issue** — `clEnqueueReadBuffer` is blocking, so any kernel-completion race would surface as garbage that varies in magnitude, which is not what we see.
- **Not a CPU-side bug** — Phase 2a confirms `CpuEngine.TensorMatMul` agrees with the hand-written reference to `max_abs_err = 5.96E-07`.

## Likely root cause hypotheses (for triage)

1. **AMD RDNA1 driver bug specific to `gfx1012`** — this is an underserved consumer GPU; the OpenCL kernel JIT-compile may produce a buggy SPIR-V/AMDGCN binary on this exact arch. Could be tested by another `gfx10xx` GPU (5700XT, 6700XT) — if those work, narrow to gfx1012.
2. **Kernel arg packing for `__kernel void gemm_small`** — the kernel source `for (k=0; k<K; k++) sum = fma(A[row*K+k], B[k*N+col], sum);` is correct, so the bug is upstream of the kernel (host-side arg setup). Possible: `kernel.SetArg(3, M)` / `(4, N)` / `(5, K)` are being passed in the wrong order, or `alpha`/`beta` are being uploaded as wrong types (int vs float). The values being roughly the right magnitude (just 3x smaller) is suspicious — would happen if `K=21` got passed instead of `K=64` (which would explain 1/3 accumulation, 1/3 magnitude).
3. **Buffer upload row-major vs column-major mismatch** — the upload via `backend.AllocateBuffer(floatData)` may be using a layout the kernel doesn't expect. CPU uses row-major; if the kernel expects column-major (e.g., FORTRAN convention) and reads `A[row*K+k]`, the resulting matrix is computed against a transposed input. But the values being smaller in magnitude (not just permuted) argues against this.

The 3x magnitude difference makes hypothesis 2 the most likely culprit (`K` being passed as a fraction of its true value).

## Severity

P0 against the `AiModelBuilder.ConfigureGpuAcceleration()` user contract. Any HE-consumer who turns on GPU acceleration on RDNA1 hardware will get a model that trains on incorrect gradients (silently). The `Aggressive` usage level explicitly promises maximum speedup with no fallback — that fails very loudly here.

## Workaround (HE side, today)

- Default `useGpu=false` on this host.
- `AIDOTNET_DIRECTGPU_BACKENDS=cpu` env var to force CPU through the module-init auto-detect.
- Call `AiDotNetEngine.ResetToCpu()` in test setup before any tensor op.

## Reference

- Test: `tests/Learning/Phase_PAPER_A_PathB_GpuAcceleration_Sanity.cs` in `ooples/HarmonicEngine` on branch `research/ledger-overhaul-and-l4-reification`
- Mechanism analysis: `docs/ledger/mechanism_failures.md` ("GPU acceleration diagnostic audit + DirectGpu OpenCL kernel correctness failure — Phase PAPER-A Path B (2026-05-17)")

**Comment by ooples** [comment]
## Update: native CLBlast (`AiDotNet.Native.CLBlast`) UNBLOCKS the ≥128 path; only the in-tree `gemm_small` (<128) remains broken

Installing **`AiDotNet.Native.CLBlast` 0.82.5** (the native library — previously `HasClBlast=False`, so only the in-tree `.cl`-compiled kernels were in play) on the same host (AMD gfx1012/RDNA1, RX 5500 XT, Win 10.0.26220, .NET 10) changes the picture decisively. Re-running float matmul GPU-vs-CPU across sizes:

| size | max_abs_err vs CPU | outcome |
|---|---|---|
| 64×64   | ~1e+34 (NaN/overflow) | **STILL WRONG** — in-tree `gemm_small` path |
| 128×128 | 1.91e-6 | **CORRECT** (native CLBlast) |
| 256×256 | 3.22e-6 | **CORRECT** |
| 512×512 | 7.15e-6 | **CORRECT** |

So the root cause is confirmed to be **the in-tree OpenCL kernels**, not the GPU/driver: with native CLBlast present, the dispatcher routes M,N≥128 through CLBlast and gets float-correct results (~1e-6), while M<128 (or N<128) still falls to the broken in-tree `gemm_small` and returns garbage/NaN.

### Suggested fixes
1. **Lower/remove the small-matrix threshold** that sends <128 matmuls to in-tree `gemm_small` when native CLBlast is available — route all sizes through CLBlast.
2. Or fix the `gemm_small` host-side arg packing (the earlier 3× magnitude on 64×64 pointed at a `K` arg mis-pack; it now NaNs with CLBlast loaded, suggesting a buffer/dispatch interaction).
3. Recommend `AiDotNet.Native.CLBlast` as a **required** companion for `ConfigureGpuAcceleration()` on OpenCL hosts — without it, the GPU path is silently wrong.

### Practical takeaway for consumers
On RDNA1, GPU acceleration IS usable for real (≥128-dim) workloads **if** native CLBlast is installed; keep all matmul dims ≥128 (or force CLBlast for all sizes) to avoid the broken `gemm_small` path.

Repro: `tests/Learning/Phase_GpuClblast_Probe.cs` in `ooples/HarmonicEngine` (branch `research/ledger-overhaul-and-l4-reification`).

**Comment by ooples** [comment]
## Update: training-side impact confirmed — facade Transformer training on GPU is CORRUPTED (CPU control)

Beyond the standalone-matmul correctness (fixed for ≥128 by native CLBlast, still broken for <128 `gemm_small`),
the **training** path is corrupted on this host (gfx1012/RDNA1, RX 5500 XT) because a Transformer's **per-head
attention matmuls are <128** (head_dim, and short sequences), so they hit the broken `gemm_small`/in-tree path.

Identical facade config (`AiDotNetFacadeTransformerLMPredictor`, AiModelBuilder + ConfigureModel/Optimizer/
DataLoader + BuildAsync; d_model=128, heads, layers, byte-LM V=256), trained on the same corpus, next-byte top1:

| engine | size | top1 | wall |
|---|---|---|---|
| **GPU** (DirectGpu+CLBlast) | 50K | **0.00%** | 850s |
| **CPU** (managed AVX2) | 25K–200K | **16.00%** (unigram-mode baseline) | 139–2013s |

The GPU result (0.00%) is *below* even the unigram baseline the CPU reaches — i.e., GPU training produces a
model worse than "always predict the most common byte". Combined with the standalone-matmul evidence, the
attention path's <128 matmuls corrupt the forward/backward on GPU. (The 850s for a 50K-byte build is also
pathologically slow vs CPU's 139s for 25K.)

**Consequence:** `ConfigureGpuAcceleration()` silently produces a broken model for ANY Transformer on RDNA1,
even with native CLBlast installed, because attention dims are intrinsically <128. The ≥128 CLBlast fix only
helps pure ≥128-dim MLPs. **Fix priority:** route the <128 attention matmuls through CLBlast too (pad to 128 /
batch heads) or fix `gemm_small`, otherwise GPU Transformer training cannot be trusted on this arch.

Repro: `tests/Learning/Phase_FacadeTransformer_ScalingExponent.cs` (GPU vs `AiDotNetEngine.ResetToCpu()`) in
`ooples/HarmonicEngine` branch `research/ledger-overhaul-and-l4-reification`.
